src/cryptographic_primitives/proofs/sigma_correct_homomorphic_elgamal_enc.rs

```diff
@@ -24,11 +23,12 @@ use super::ProofError;
 ///
 #[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
 #[serde(bound = "")]
-pub struct HomoELGamalProof<E: Curve> {
+pub struct HomoELGamalProof<E: Curve, H: Digest + Clone> {
     pub T: Point<E>,
     pub A3: Point<E>,
     pub z1: Scalar<E>,
     pub z2: Scalar<E>,
+    _ph: PhantomData<H>,
```

Comment:
Having `PhantomData<H>` as a field of a structure makes the structure `Send`able and `Sync`able only if [`H: Send`][Send] and [`H: Sync`][Sync] respectively, whereas H being or not Send or Sync doesn't actually affect structure ability to be sent or synced. It makes a lot of pain when dealing with multi-thread code.

We have several options:
1) Use external library like [phantom-type](https://crates.io/crates/phantom-type)
2) Use `PhantomData<fn() -> H>` which is always `Send + Sync`
3) Be aware of and ignore it 🙂

[Send]: https://doc.rust-lang.org/std/marker/struct.PhantomData.html#impl-Send
[Sync]: https://doc.rust-lang.org/std/marker/struct.PhantomData.html#impl-Sync

Reply:
Also we can take advantage and create own "marker" `HashChoice<H>` that will serialize into `"hash_type": "sha256"`, and deserialize only if `"hash_type"` matches choice of hash. It will add some verbosity to the serialized proof, but adds communication size, so I guess it's not really an option.

Reply:
do we do that for the curve selection here?

and I like `PhantomData<fn() -> H>` I'll need to re-read the variance rules to make sure I understand all the implications but sounds good

Reply:
> do we do that for the curve selection here?

Do you mean if we using `PhantomData` for `E` generics? Never had to (yet?). Every structure that depends on choice of curve `E` stores either `Point<E>` or `Scalar<E>` as a field (or stores another structure that stores point/scalar as a field), so `E` is always used

Reply:
> Do you mean if we using PhantomData for E generics?

No, I meant serializing the curve name and only de-serializing if it's the correct curve name.

> so E is always used 

Well there are 2 places where this isn't correct and we use `PhantomData` for the curve:
https://github.com/ZenGo-X/curv/blob/02d683dd5d59a3a22c8743e1cda8b5a81c9903b9/src/cryptographic_primitives/hashing/merkle_tree.rs#L24
https://github.com/ZenGo-X/curv/blob/02d683dd5d59a3a22c8743e1cda8b5a81c9903b9/src/cryptographic_primitives/commitments/pedersen_commitment.rs#L21

Reply:
> No, I meant serializing the curve name and only de-serializing if it's the correct curve name.

Yes, every time Point/Scalar are deserialized, we do curve check:
https://github.com/ZenGo-X/curv/blob/02d683dd5d59a3a22c8743e1cda8b5a81c9903b9/src/elliptic/curves/wrappers/serde_support.rs#L49
https://github.com/ZenGo-X/curv/blob/02d683dd5d59a3a22c8743e1cda8b5a81c9903b9/src/elliptic/curves/wrappers/serde_support.rs#L87

Reply:
> Well there are 2 places where this isn't correct and we use PhantomData for the curve:

Uh forgot about those places. I believe merkle_tree should be updated to store `PhantomData<fn() -> E>` too. No need to touch PedersenCommitment, it has only static methods.

Reply:
Note that also Generator isn't necessarily Sync/Send
https://github.com/ZenGo-X/curv/blob/master/src/elliptic/curves/wrappers/generator.rs#L44

Reply:
Generator has the same autotraits implemented as `&Point<E>` does, that was done on purpose to match implicitly expected behaviour. I agree that it's not really required, but seems fine to me.